From: Fiona Klute <fiona.klute@gmx.de>
Date: Thu, 27 Aug 2026 17:10:35 +0200
Subject: fix(busybox): do not rely on CONFIG_FEATURE_INSTALLER=y

busybox --list-full provides full paths only if busybox was built with
CONFIG_FEATURE_INSTALLER=y [1], otherwise the result is only a list of applet
names. Using those plain applet names, links will be placed in / in the
initramfs, which is not on PATH, causing chaos.

If we get plain applet names, try to find their installed paths, and use those
if found. Otherwise guess usr/bin.

[1] https://github.com/vda-linux/busybox_mirror/blob/15666f3e9a0b54305cec1b01f34bc5b2a1794e74/libbb/appletlib.c#L857-L860

Fixes: fcf3abd126a2 ("fix(busybox): use full path provided by busybox --list-all")
Origin: upstream, https://github.com/dracut-ng/dracut/commit/98306da0ce9b00a634896d3095aee75eaf0fc06a
---
 modules.d/81busybox/module-setup.sh | 16 ++++++++++++++--
 1 file changed, 14 insertions(+), 2 deletions(-)

diff --git a/modules.d/81busybox/module-setup.sh b/modules.d/81busybox/module-setup.sh
index 7bce09e..03a3c2b 100755
--- a/modules.d/81busybox/module-setup.sh
+++ b/modules.d/81busybox/module-setup.sh
@@ -13,15 +13,27 @@ check() {
 
 # called by dracut
 install() {
-    local _path _busybox _busybox_path
+    local _path _p _busybox _busybox_path
     local _dstdir="${dstdir:-"$initdir"}"
     local _progs=()
     _busybox=$(find_binary busybox)
     _busybox_path="/usr/bin/busybox"
 
-    # do not depend on CONFIG_FEATURE_INSTALLER
     # install busybox symlinks manually
     for _path in $($_busybox --list-full); do
+        # if busybox is built without CONFIG_FEATURE_INSTALLER=y, the list
+        # has only plain names
+        if [[ ${_path##*/} == "${_path}" ]]; then
+            _p=$(find_binary "${_path}")
+            if [[ $_p ]]; then
+                # if the applet is installed, use its path
+                _path="${_p}"
+            else
+                # otherwise guess usr/bin/
+                _path="usr/bin/${_path}"
+            fi
+        fi
+
         if [[ ${_path##*/} == busybox ]]; then
             _busybox_path="/${_path#/}"
         else
